### PR TITLE
Log in

### PR DESCRIPTION
**Pull request by Asman97**

### CODE REVIEW

**Comment by dEisenbeiss** [comment]
for the next time, try to improve your commit messages :)

- should always be english
- should contain usefull information about the changes [added authentication service / removed file xy / fixed a bug]
- should include a reference to the user story that is worked on [like: #1, #99, or even NO-ISSUE when there was no related story]

e.g.: 
"#1 added necessary files for spring application"

**Comment by Asman97** [comment]
> for the next time, try to improve your commit messages :)
> 
> * should always be english
> * should contain usefull information about the changes [added authentication service / removed file xy / fixed a bug]
> * should include a reference to the user story that is worked on [like: [Spring MVC Projekt erstellen #1](https://github.com/IoT-Ausbildung/grade-calculator/issues/1), #99, or even NO-ISSUE when there was no related story]
> 
> e.g.: "#1 added necessary files for spring application"

**Comment by Asman97** [comment]
Yup

**Review by dEisenbeiss — COMMENTED** [review]
- login does not have validation messages anymore
- login expects an email but the description of the field is username
- login: input fields have different sizes, just make them use the available space
- index: i would remove the "Hello anonymousUser!" message when not authenticated

File: src/main/java/com/example/gradecalculator/web/controller/UserController.java [hunk 1]
```diff
@@ -60,14 +70,40 @@ private String registerPost(Model model, @Valid @ModelAttribute UserSignUpTO reg
...
         }
 
         var userTypes = userTypeRepository.findAll();
         model.addAttribute("userTypes", userTypes);
         model.addAttribute("registration", registration);
         model.addAttribute("itemErrors", errors);
-        return "register";
+        return "/register";
+    }
+/*    @GetMapping("/login")
```
Comment (dEisenbeiss) [inline src/main/java/com/example/gradecalculator/web/controller/UserController.java:82]: remove unused code

Comment (Asman97) [reply src/main/java/com/example/gradecalculator/web/controller/UserController.java]: done

File: src/main/java/com/example/gradecalculator/model/MyUserDetails.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+import com.example.gradecalculator.entities.User;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
+
+import java.util.Arrays;
+import java.util.Collection;
+
+public class MyUserDetails implements UserDetails {
```
[the remark this answers is not in the document]
Comment (Asman97) [reply src/main/java/com/example/gradecalculator/model/MyUserDetails.java:11]: done

[the remark this answers is not in the document]
Comment (dEisenbeiss) [reply src/main/java/com/example/gradecalculator/model/MyUserDetails.java:11]: i see no changes

File: src/main/java/com/example/gradecalculator/web/model/LoginDTO.java
Comment (dEisenbeiss) [inline src/main/java/com/example/gradecalculator/web/model/LoginDTO.java:1]: this can be removed, as we do not have a custom login anymore, right?

File: src/main/resources/application.properties [hunk 3]
```diff
@@ -8,4 +8,5 @@ server.error.path=/error
 spring-jpa.properties.hibernate.dialect=org.hibernate.dialect.PostgreSQLDialect
 spring-jpa.properties.hibernate.format_sql=true
 server.port=8080
-
+spring.security.user.name=dianamirlan@gmail.com
```
Comment (dEisenbeiss) [inline src/main/resources/application.properties:11]: there should never be personal information in a repository

also this is not used as I would suppose, so it should be removed

File: src/main/resources/templates/login.html [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    <script>
+        $(document).ready(function () {
+            $("#nav-placeholder").load("/navbar")
+        })
+    </script>
+</head>
+<body>
+<div id="nav-placeholder"></div>
+<div style="display: flex; justify-content: center; margin-top: 90px;">
+    <form th:action="@{/login}" method="post" style="width: 600px; padding: 20px; border: 1px solid #ccc; border-radius: 10px;">
```
Comment (dEisenbeiss) [inline src/main/resources/templates/login.html:32]: inline style shoud be avoided
better approach would be to give the tag a class and define the style in your css files

Comment (dEisenbeiss) [reply src/main/resources/templates/login.html:32]: see also for every other style attribute in html

File: src/main/java/com/example/gradecalculator/model/MyUserDetails.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Asman97) [reply src/main/java/com/example/gradecalculator/model/MyUserDetails.java]: sorry it turns out I haven't pushed it

File: src/main/java/com/example/gradecalculator/web/model/LoginDTO.java
Comment (Asman97) [reply src/main/java/com/example/gradecalculator/web/model/LoginDTO.java:1]: yes that's right

File: src/main/java/com/example/gradecalculator/config/PasswordConfig.java [hunk 5]
```diff
@@ -2,14 +2,15 @@
...
 
     @Bean
     public PasswordEncoder passwordEncoder() {
-        return new BCryptPasswordEncoder();
+        return PasswordEncoderFactories.createDelegatingPasswordEncoder();
     }
-}
\ No newline at end of file
+
+    }
```
Comment (dinowol) [inline src/main/java/com/example/gradecalculator/config/PasswordConfig.java:16]: Why is there a new line (more spacing) but also worse bracket placement if this did not gave to change at all?

File: src/main/java/com/example/gradecalculator/repository/GradeTypeRepository.java [hunk 6]
```diff
@@ -6,7 +6,6 @@
 
 import com.example.gradecalculator.entities.GradeSystem;
 import com.example.gradecalculator.entities.GradeType;
-import org.springframework.data.repository.CrudRepository;
```
Comment (dinowol) [inline src/main/java/com/example/gradecalculator/repository/GradeTypeRepository.java:9]: can you really just delete this import?

File: src/main/java/com/example/gradecalculator/service/UserDetailsServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+@Service
+public class UserDetailsServiceImpl implements UserDetailsService {
+
+    @Autowired
+    private UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String userName)
+            throws UsernameNotFoundException {
```
Comment (dinowol) [inline src/main/java/com/example/gradecalculator/service/UserDetailsServiceImpl.java:20]: please mind indentation. it is very hard to say where this bracket closes again

File: src/main/java/com/example/gradecalculator/service/UserService.java [hunk 8]
```diff
@@ -38,4 +44,6 @@ public User createUser(UserSignUpTO registration) {
         userRepository.save(user);
         return user;
     }
+
```
Comment (dinowol) [inline src/main/java/com/example/gradecalculator/service/UserService.java:47]: ???

File: src/main/java/com/example/gradecalculator/web/model/LoginDTO.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
+package com.example.gradecalculator.web.model;
+
+import lombok.Getter;
+
+@Getter
+public class LoginDTO {
+    private String email;
+    private String password;
+
+    public LoginDTO() {
```
Comment (dinowol) [inline src/main/java/com/example/gradecalculator/web/model/LoginDTO.java:10]: ?

File: src/main/resources/static/css/style.css [hunk 10]
```diff
@@ -60,4 +59,13 @@ h1 {
 }
 .login {
     --bs-btn-hover-bg: 0.3s;
-}
\ No newline at end of file
+}
+
+.email {
+    margin-top: 50px ;
+}
```
Comment (dinowol) [inline src/main/resources/static/css/style.css:66]: you really like to add empty lines in the java code where they do not belong but do not put them in css, where they should be so stuff remains readable ^^

File: src/main/resources/templates/index.html [hunk 11]
```diff
@@ -22,13 +23,15 @@
...
         })
     </script>
 </head>
-
 <body>
-  <div id="nav-placeholder"></div>
+<div>
+    <div id="nav-placeholder"></div>
+</div>
+<h1 th:if="${#authorization.expression('isAuthenticated()')}" th:inline="text" style="text-align: center; margin-top: 90px">Hello <span th:remove="tag" sec:authentication="name"></span>!</h1>
```
Comment (dinowol) [inline src/main/resources/templates/index.html:35]: style belongs in css

**Review by dinowol — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/gradecalculator/config/PasswordConfig.java [hunk 12]
```diff
@@ -2,14 +2,15 @@
...
+import org.springframework.security.crypto.factory.PasswordEncoderFactories;
 import org.springframework.security.crypto.password.PasswordEncoder;
 
 @Configuration
 public class PasswordConfig {
 
     @Bean
     public PasswordEncoder passwordEncoder() {
-        return new BCryptPasswordEncoder();
+        return PasswordEncoderFactories.createDelegatingPasswordEncoder();
```
Comment (dEisenbeiss) [inline src/main/java/com/example/gradecalculator/config/PasswordConfig.java:13]: Why did you change the password encoder?

File: src/main/java/com/example/gradecalculator/config/WebSecurityConfig.java [hunk 13]
```diff
@@ -1,26 +1,68 @@
...
+                        .permitAll()
+                );
         return http.build();
     }
-
     @Bean
     public WebSecurityCustomizer webSecurityCustomizer() throws Exception {
         return (web) -> web.ignoring().requestMatchers("/resources/**");
     }
+    private final AuthenticationConfiguration authConfiguration;
```
Comment (dEisenbeiss) [inline src/main/java/com/example/gradecalculator/config/WebSecurityConfig.java:54]: fields should be at the top of the class to improve readablity of the file.

There are multiple conventions that differ from each other, e.g:
https://web.archive.org/web/20130516014426/http://www.oracle.com/technetwork/java/codeconventions-141855.html#1852

You should pick one convention to use in your project and stick to it for all implementations

File: src/main/java/com/example/gradecalculator/config/WebSecurityConfig.java [hunk 14]
```diff
@@ -1,26 +1,68 @@
...
 @EnableWebSecurity
 public class WebSecurityConfig {
 
+    private final PasswordConfig passwordConfig;
+
+    @Bean
+    public UserDetailsService userDetailsService() {
+        return new UserDetailsServiceImpl();
+    }
+    public WebSecurityConfig(
```
Comment (dEisenbeiss) [inline src/main/java/com/example/gradecalculator/config/WebSecurityConfig.java:28]: ordering in file: constructor should come before methods

File: src/main/java/com/example/gradecalculator/config/WebSecurityConfig.java [hunk 15]
```diff
@@ -1,26 +1,68 @@
...
     public WebSecurityCustomizer webSecurityCustomizer() throws Exception {
         return (web) -> web.ignoring().requestMatchers("/resources/**");
     }
+    private final AuthenticationConfiguration authConfiguration;
 
+    @Bean
+    public AuthenticationManager authenticationManager() throws Exception {
+        return authConfiguration.getAuthenticationManager();
+    }
+    @Bean
```
Comment (dEisenbeiss) [inline src/main/java/com/example/gradecalculator/config/WebSecurityConfig.java:60]: add a line between methods to have consistent spacing

File: src/main/java/com/example/gradecalculator/repository/UserRepository.java [hunk 16]
```diff
@@ -4,13 +4,20 @@
...
+@Repository
 public interface UserRepository extends CrudRepository<User, Long> {
+
     @Query("SELECT CASE WHEN COUNT(u) > 0 THEN true ELSE false END FROM User u WHERE u.email = :email")
     boolean existsByEmail(@Param("email") String email);
+
     @Query("SELECT CASE WHEN COUNT(u) > 0 THEN true ELSE false END FROM User u WHERE u.userName = :userName")
     boolean existsByUserName(@Param("userName") String userName);
+    @Query("SELECT u FROM User u WHERE u.email = :email AND u.encodedPassword = :password")
+    User findByEmailAndPassword(@Param("email") String email, @Param("password") String password);
```
Comment (dEisenbeiss) [inline src/main/java/com/example/gradecalculator/repository/UserRepository.java:18]: this method is not used and should be removed

File: src/main/java/com/example/gradecalculator/service/UserService.java [hunk 17]
```diff
@@ -4,14 +4,20 @@
...
 
 @Service
 public class UserService {
-
+    public String loginUser(LoginDTO loginDTO) {
+        return null;
+    }
+    public boolean isValidUser(String email, String password) {
+        return true;
+    }
```
Comment (dEisenbeiss) [inline src/main/java/com/example/gradecalculator/service/UserService.java:20]: those methods have no usage

File: src/main/java/com/example/gradecalculator/web/controller/UserController.java [hunk 18]
```diff
@@ -26,16 +30,22 @@ public class UserController {
 
     private final UserRepository userRepository;
     private final UserTypeRepository userTypeRepository;
+    private final AuthenticationManager authenticationManager;
```
Comment (dEisenbeiss) [inline src/main/java/com/example/gradecalculator/web/controller/UserController.java:33]: this field is never used

File: src/main/java/com/example/gradecalculator/web/model/LoginDTO.java [hunk 19]
```diff
@@ -0,0 +1,40 @@
+package com.example.gradecalculator.web.model;
+
+import lombok.Getter;
+
+@Getter
+public class LoginDTO {
```
Comment (dEisenbeiss) [inline src/main/java/com/example/gradecalculator/web/model/LoginDTO.java:6]: this class is not used anymore, as the custom login implementation was replaced with the built in one.
you should remove this class

**Review by dEisenbeiss — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/gradecalculator/repository/GradeTypeRepository.java (hunk 6 above)
Comment (Asman97) [reply src/main/java/com/example/gradecalculator/repository/GradeTypeRepository.java:9]: I can't, its being used for repository

File: src/main/java/com/example/gradecalculator/web/controller/UserController.java (hunk 18 above)
Comment (dEisenbeiss) [reply src/main/java/com/example/gradecalculator/web/controller/UserController.java:33]: yes, but its only for initializing the field; this should therefore also be removed

Comment (dEisenbeiss) [reply src/main/java/com/example/gradecalculator/web/controller/UserController.java:33]: you can quickly see this in the IDE as the field name is grey:
image


the message in the IDE is: "Private field 'authenticationManager' is assigned but never accessed"

**Comment by dEisenbeiss** [comment]
As discussed in the daily: login via username should also be possible

**Comment by dEisenbeiss** [comment]
image

this should match

**Comment by dEisenbeiss** [comment]
image

there is a lot of empty space, that could be reduced

**Comment by Asman97** [comment]
> image
> 
> there is a lot of empty space, that could be reduced

I couldn't understand why its being so much space between username and email. Any ideas? Margin top I can reduce

**Review by dEisenbeiss — APPROVED** [review]

**Review by dinowol — APPROVED** [review]
